query-parser/src/parse.rs

```diff
@@ -93,34 +96,73 @@ impl<I> Query<I>
 
 def_value_satisfy_parser_fn!(Query, variable, Variable, Variable::from_value);
 def_value_satisfy_parser_fn!(Query, source_var, SrcVar, SrcVar::from_value);
+def_value_satisfy_parser_fn!(Query, predicate_fn, PredicateFn, PredicateFn::from_value);
+def_value_satisfy_parser_fn!(Query, fn_arg, FnArg, FnArg::from_value);
 
 pub struct Where<I>(::std::marker::PhantomData<fn(I) -> I>);
 
-def_value_satisfy_parser_fn!(Where, pattern_value_place, PatternValuePlace, PatternValuePlace::from_value);
-def_value_satisfy_parser_fn!(Where, pattern_non_value_place, PatternNonValuePlace, PatternNonValuePlace::from_value);
+def_value_satisfy_parser_fn!(Where,
+                             pattern_value_place,
+                             PatternValuePlace,
+                             PatternValuePlace::from_value);
+def_value_satisfy_parser_fn!(Where,
+                             pattern_non_value_place,
+                             PatternNonValuePlace,
+                             PatternNonValuePlace::from_value);
+
+/// Take a vector Value containing one vector Value, and return the `Vec` inside the inner vector.
+fn unwrap_nested(x: edn::Value) -> Option<Vec<edn::Value>> {
+    match x {
+        edn::Value::Vector(mut v) => {
+            match v.pop() {
+                Some(edn::Value::Vector(items)) => Some(items),
```

Comment:
You're returning consuming your inputs and returning owned values; can't you match `Some(edn::Value::List(linked_list))` and produce a `Vec` out of it?